### PR TITLE
Support setting backend for machine mem

### PR DESCRIPTION
**Pull request by qingwangrh**
ID:2110836
Signed-off-by: qingwangrh <qinwang@redhat.com>

### CODE REVIEW

**Comment by qingwangrh** [comment]
@luckyh @YongxueHong Please help to review. Thanks.

**Comment by qingwangrh** [comment]
Passed key definition in cfg file.

backend_mem = memory-backend-memfd
vm_mem_share = yes

File: virttest/qemu_capabilities.py [hunk 1]
```diff
@@ -22,6 +22,8 @@ class Flags(object):
     SMP_DIES = _auto_value()
     INCOMING_DEFER = _auto_value()
     MIGRATION_PARAMS = _auto_value()
+    DAEMONIZE = _auto_value()
+    PIDFILE = _auto_value()
```
Comment (luckyh) [inline virttest/qemu_capabilities.py:26]: Hi @qingwangrh would you please elaborate on the purpose of adding these two new flags?

File: virttest/qemu_vm.py [hunk 2]
```diff
@@ -1011,6 +1011,8 @@ def add_memorys(devices, params):
                 name = output[0]
                 backend_options = {}
                 backend_options["size_mem"] = "%sM" % params["mem"]
+                if params.get("backend_mem"):
```
Comment (luckyh) [inline virttest/qemu_vm.py:1014]: For having a consistent design with other machine-level mem relevant params (like `vm_mem_share`) I'd suggest using the name `vm_mem_backend` for the param instead.

File: virttest/qemu_capabilities.py (hunk 1 above)
Comment (qingwangrh) [reply virttest/qemu_capabilities.py:26]: The QSD is keep evolving ,the options are supported in different version.
I hope QSD may share the capablitlies definition with qemu. So i add them here.

File: virttest/qemu_vm.py (hunk 2 above)
Comment (qingwangrh) [reply virttest/qemu_vm.py:1014]: "backend_mem" is the key in using. if i use an new key does it mean all relevant cases need to be changed?

File: virttest/qemu_capabilities.py (hunk 1 above)
Comment (qingwangrh) [reply virttest/qemu_capabilities.py:26]: The qemu also include relevant options. I can put them in tp-qemu depend on your decision. 

root@qing /home/kar/internal_cfg/guest-os $ /usr/libexec/qemu-kvm  -h|grep daem
-daemonize      daemonize QEMU after initializing
root@qing /home/kar/internal_cfg/guest-os $ /usr/libexec/qemu-kvm  -h|grep pid
-pidfile file   write PID to 'file'

File: virttest/qemu_vm.py (hunk 2 above)
Comment (luckyh) [reply virttest/qemu_vm.py:1014]: > "backend_mem" is the key in using. if i use an new key does it mean all relevant cases need to be changed?

Actually, no. From my PoV, It just means we are introducing a new param to the **machine-level mem device** for specifying its backend type. Other mem devices would still use the `backend_mem` param.

File: virttest/qemu_capabilities.py (hunk 1 above)
Comment (luckyh) [reply virttest/qemu_capabilities.py:26]: > The QSD is keep evolving ,the options are supported in different version. I hope QSD may share the capablitlies definition with qemu. So i add them here.

I see, but I'd suggest making a capablitlies individual definition for the QSD binary due to the correctness concerns and maintainability reasons.

Comment (qingwangrh) [reply virttest/qemu_capabilities.py]: Updated

File: virttest/qemu_vm.py (hunk 2 above)
Comment (qingwangrh) [reply virttest/qemu_vm.py]: Agree,updated

**Comment by luckyh** [comment]
LGTM, could you please provide the log snippet about the corresponding qemu command lines? thanks in advance!

**Review by YongxueHong — APPROVED** [review]
LGTM.

**Comment by qingwangrh** [comment]
> LGTM, could you please provide the log snippet about the corresponding qemu command lines? thanks in advance!

I updated # comment 2,Please check the comment

**Comment by yanan-fu** [comment]
When i saw the patch title, the first thing i thought is setting default mem backend for memory object while i remember it use `memory-backend-ram` if no backend be specified:
https://github.com/avocado-framework/avocado-vt/blob/17fd6bd0529eabf3f59512207f24c9a89eaf17d4/virttest/qemu_devices/qcontainer.py#L2900

However, after go thought the code change, i see it is used to support setting the backend for machine mem.  
How about change the title a bit, something like: `support setting backend for machine mem`

**Comment by qingwangrh** [comment]
> When i saw the patch title, the first thing i thought is setting default mem backend for memory object while i remember it use `memory-backend-ram` if no backend be specified:
> 
> https://github.com/avocado-framework/avocado-vt/blob/17fd6bd0529eabf3f59512207f24c9a89eaf17d4/virttest/qemu_devices/qcontainer.py#L2900
> 
> However, after go thought the code change, i see it is used to support setting the backend for machine mem. How about change the title a bit, something like: `support setting backend for machine mem`

updated

**Comment by yanan-fu** [comment]
> > LGTM, could you please provide the log snippet about the corresponding qemu command lines? thanks in advance!
> 
> I updated # comment 2,Please check the comment

Call for key debug log with this patch and param `vm_mem_backend`.

**Comment by qingwangrh** [comment]
> > > LGTM, could you please provide the log snippet about the corresponding qemu command lines? thanks in advance!
> > 
> > 
> > I updated # comment 2,Please check the comment
> 
> Call for key debug log with this patch and param `vm_mem_backend`.

Passed key definition in cfg file.

vm_mem_backend = memory-backend-memfd
vm_mem_share = yes

It get command line like following:
......
-machine q35,memory-backend=mem-machine_mem
-device pcie-root-port,id=pcie-root-port-0,multifunction=on,bus=pcie.0,addr=0x1,chassis=1
-device pcie-pci-bridge,id=pcie-pci-bridge-0,addr=0x0,bus=pcie-root-port-0
-nodefaults
-device VGA,bus=pcie.0,addr=0x2
-m 12288
-object memory-backend-memfd,size=12288M,share=yes,id=mem-machine_mem \

**Review by yanan-fu — APPROVED** [review]

**Review by luckyh — APPROVED** [review]

**Comment by luckyh** [comment]
Thanks to all.

**Pull request merged**